### ISSUE TITLE
write to stdout fails under PY3

### ISSUE THREAD

**Issue by alimanfoo** [alimanfoo/petl#308]
Writing to stdout fails under PY3 because petl treats as binary stream and tries to write bytes.

**Comment by krisztianfekete** [comment]
stdin/stdout functionality is high priority for me (pipes, with processes on different cores!).

short answer:
```python
sys.stdout.flush()
sys.stdout.buffer.write(some_binary_data)
```

http://nullege.com/codes/search/sys.stdout.buffer.write

Same answer from different angles (from first page of https://www.google.com/search?q=python3+sys.stdout+binary):

http://stackoverflow.com/questions/908331/how-to-write-binary-data-in-stdout-in-python-3
http://bugs.python.org/issue18512
https://wiki.python.org/moin/PortingPythonToPy3k
https://docs.python.org/dev/library/io.html?highlight=buffer#io.TextIOBase.buffer

## stdin

It is fair to assume the same problem for stdin as well (Python3: text mode), people are working around it as well:

http://nullege.com/codes/search/sys.stdin.buffer.read

**Comment by krisztianfekete** [comment]
It is quite crazy, but the above solution is officially marked unportable: https://docs.python.org/dev/library/io.html?#io.TextIOBase.buffer

http://bugs.python.org/msg195593 from http://bugs.python.org/issue18512 has:
```python
sys.stdout.detach()
```
which is also used by some people (http://nullege.com/codes/search/sys.stdout.detach)

Unfortunately it has the same portability warning: https://docs.python.org/dev/library/io.html?highlight=buffer#io.TextIOBase.detach

**Comment by alimanfoo** [comment]
Hi @krisztianfekete, thank you for these comments, they are extremely helpful.

Another option may be to re-open the underlying file descriptor to obtain a binary stream, i.e., os.fdopen(sys.stdout.fileno()). Of course sys.stdout may have been replaced by something other than the original stdout, so there is no guarantee that it will support either fileno() or buffer or detach(). Some sort of duck-type checking may be the only answer. I'm looking into this at the moment, but feel free to post any suggestions.

**Comment by krisztianfekete** [comment]
@alimanfoo I am quite confident that "not part of the API" comment can be ignored.  It is documented as a the way to go in a note here: https://docs.python.org/3/library/sys.html#sys.stdin
and also appears in code fragment here: https://docs.python.org/3/library/sys.html#sys.displayhook

**Comment by alimanfoo** [comment]
This is the heuristic for obtaining a stdout binary stream implemented in v1.0.1: https://github.com/alimanfoo/petl/blob/issue_308/petl/io/sources.py#L101

Passes tests under Windows and Linux, py26/py27/py34, but any comments welcome.

**Comment by alimanfoo** [comment]
Btw tests just verify that writing to stdout does not raise an exception.
